Поиск инсайтов в данных продуктовых магазинов компании. В файле содержится информация о покупках людей.  
Необходимо выяснить, какие пары товаров пользователи чаще всего покупают вместе. По сути, необходимо найти паттерны покупок, что позволит оптимизировать размещение продуктов в магазине для удобства пользователей и увеличения выручки.       
_________
**Описание данных:**                  
 id – означает покупку (в одну покупку входят все товары, купленные пользователем во время 1 похода в магазин)      
 Товар – наименование товара      
 Количество – число единиц купленного товара        

In [1]:
import pandas as pd
from tqdm.auto import tqdm
import scipy.special

In [2]:
data = pd.read_csv('https://stepik.org/media/attachments/lesson/409319/test1_completed.csv')
data.head(5)

id            Товар  Количество
0  17119            Лимон         1.1
1  17119  Лимон оранжевый         0.7
2  17119        Лук-порей        10.0
3  17119     Лук репчатый         2.5
4  17119    Малина свежая         1.0

In [3]:
# Число уникальных товаров:
data['Товар'].nunique()

199

In [4]:
# Список всех товаров:
products = list(data['Товар'].unique())
products

['Лимон',
 'Лимон оранжевый',
 'Лук-порей',
 'Лук репчатый',
 'Малина свежая',
 'Морковь немытая',
 'Черешня сушеная',
 'Изюм Султана',
 'Капуста цветная',
 'Бразильский орех',
 'Дыня Гуляби',
 'Шпинат мини',
 'Изюм Голд',
 'Огурцы Луховицкие',
 'Чеснок весовой',
 'Огурцы Белоруссия',
 'Черешня отборная',
 'Абрикосы молдавские',
 'Кукуруза мини',
 'Грибы лисички',
 'Слива Президент',
 'Апельсины столовые',
 'Кукуруза свежая',
 'Салат Руккола',
 'Гранат',
 'Подпись открытки',
 'Лук зеленый',
 'Цукаты (брусочки)',
 'Кокос питьевой',
 'Петрушка',
 'Помидоры Парадайз',
 'Ягоды Годжи',
 'Укроп',
 'Черника свежая',
 'Персики',
 'Арбуз',
 'Салат Айсберг',
 'Галангал корень',
 'Петрушка Кудрявая',
 'Яблоки Гала',
 'Курага коричневая',
 'Горох стручковый',
 'Персики узбекские',
 'Персик сушеный',
 'Груша Форел',
 'Яблоки Айдаред',
 'Яблоки Симиренко',
 'Кабачки',
 'Кинза',
 'Морковь мытая',
 'Капуста Кольраби',
 'Груша Аббат',
 'Шпинат',
 'Вишня садовая',
 'Сахар-песок, 5кг',
 'Сельдерей корень

In [5]:
# Составим список всех возможных пар товаров:
pairs = []
for i in range(len(products)):
    for j in range(i+1, len(products)):
        pairs.append([products[i],products[j]])
pairs

[['Лимон', 'Лимон оранжевый'],
 ['Лимон', 'Лук-порей'],
 ['Лимон', 'Лук репчатый'],
 ['Лимон', 'Малина свежая'],
 ['Лимон', 'Морковь немытая'],
 ['Лимон', 'Черешня сушеная'],
 ['Лимон', 'Изюм Султана'],
 ['Лимон', 'Капуста цветная'],
 ['Лимон', 'Бразильский орех'],
 ['Лимон', 'Дыня Гуляби'],
 ['Лимон', 'Шпинат мини'],
 ['Лимон', 'Изюм Голд'],
 ['Лимон', 'Огурцы Луховицкие'],
 ['Лимон', 'Чеснок весовой'],
 ['Лимон', 'Огурцы Белоруссия'],
 ['Лимон', 'Черешня отборная'],
 ['Лимон', 'Абрикосы молдавские'],
 ['Лимон', 'Кукуруза мини'],
 ['Лимон', 'Грибы лисички'],
 ['Лимон', 'Слива Президент'],
 ['Лимон', 'Апельсины столовые'],
 ['Лимон', 'Кукуруза свежая'],
 ['Лимон', 'Салат Руккола'],
 ['Лимон', 'Гранат'],
 ['Лимон', 'Подпись открытки'],
 ['Лимон', 'Лук зеленый'],
 ['Лимон', 'Цукаты (брусочки)'],
 ['Лимон', 'Кокос питьевой'],
 ['Лимон', 'Петрушка'],
 ['Лимон', 'Помидоры Парадайз'],
 ['Лимон', 'Ягоды Годжи'],
 ['Лимон', 'Укроп'],
 ['Лимон', 'Черника свежая'],
 ['Лимон', 'Персики'],
 ['Лимо

In [6]:
# Число пар в списке должно быть равно числу сочетаний из n по k, где n - кол-во товаров, k = 2:
len(pairs) == scipy.special.comb(data['Товар'].nunique(), 2)

True

Другими словами, из 199 имеющихся товаров мы составили $199! / (2! * (199-2)!) = 198 * 199 / 2 = 19 701$ всех возможных пар, чтобы далее посчитать, как часто каждая из них встречается в покупках.

In [7]:
# Преобразуем данные так, чтобы получить для каждой покупки весь список товаров:

data1 = data.groupby('id')['Товар'].apply(list).to_frame()
data1

Товар
id                                                       
17119   [Лимон, Лимон оранжевый, Лук-порей, Лук репчат...
17530   [Лимон оранжевый, Изюм Султана, Капуста цветна...
17618   [Абрикосы молдавские, Кукуруза мини, Грибы лис...
17724   [Абрикосы молдавские, Арбуз, Салат Айсберг, Га...
17814   [Арбуз, Кабачки, Кинза, Малина свежая, Морковь...
...                                                   ...
119062  [Баклажаны грунтовые, Персики, Персики армянск...
119110  [Авокадо ХАСС, Апельсины столовые, Базилик зел...
119206  [Огурцы длинноплодные, Грибы шампиньоны, Подпи...
119393  [Петрушка, Грибы Еринги, Инжир свежий, Редис, ...
119396  [Фасоль стручковая, Лук-резанец (шнитт-лук), Л...

[3273 rows x 1 columns]

In [8]:
freq = []

# Для каждого списка-пары товаров посмотрим, является ли он подмножеством множества товаров в каждой покупке, и посчитаем
# сумму таких подмножеств для всех покупок:
for i in tqdm(pairs):
    freq.append({'1_Товар':i[0],'2_Товар':i[1],'Встречаемость':data1['Товар'].apply(lambda x: set(x) >= set(i)).sum()})
    
# Теперь добавим все данные в датафрейм:    
data3 = pd.DataFrame(freq)
data3 = data3.sort_values('Встречаемость', ascending = False)
data3.head(30)

# Получаем датафрейм со встречаемостью для всех возможных пар продуктов:

1_Товар                  2_Товар  Встречаемость
2514  Огурцы Луховицкие                    Укроп            431
5338           Петрушка                    Укроп            408
2517  Огурцы Луховицкие                    Арбуз            345
2529  Огурцы Луховицкие                  Кабачки            326
5855              Укроп                    Кинза            303
4828        Лук зеленый                    Укроп            300
2511  Огурцы Луховицкие                 Петрушка            286
600        Лук репчатый        Огурцы Луховицкие            285
8240            Кабачки      Баклажаны грунтовые            284
5854              Укроп                  Кабачки            281
2545  Огурцы Луховицкие      Баклажаны грунтовые            276
5842              Укроп                    Арбуз            274
619        Лук репчатый                    Укроп            272
2512  Огурцы Луховицкие        Помидоры Парадайз            266
5870              Укроп      Баклажаны грунтовые            265
2556  Огурцы Луховицкие                   Бананы            264
2530  Огурцы Луховицкие                    Кинза            249
2572  Огурцы Луховицкие  Персики азербайджанские            244
5881              Укроп                   Бананы            239
6373              Арбуз                   Бананы            237
6346              Арбуз                  Кабачки            229
2563  Огурцы Луховицкие        Персики армянские            224
2508  Огурцы Луховицкие              Лук зеленый            221
634        Лук репчатый                  Кабачки            206
5506  Помидоры Парадайз                    Укроп            205
4825        Лук зеленый                 Петрушка            204
6389              Арбуз  Персики азербайджанские            203
5354           Петрушка                    Кинза            201
622        Лук репчатый                    Арбуз            201
6380              Арбуз        Персики армянские            200

Итак, мы получили датафрейм со встречаемостью всевозможных пар продуктов в нашем магазине. Нельзя, например, не отметить тот факт, что потребители часто покупают разную зелень вместе (Укроп-Петрушка-Кинза-Лук зеленый и т.д.) - такие комбинации встречаются очень часто, или овощи с овощами (Огурцы-Кабачки-Баклажаны-Лук репчатый-Помидоры Парадайз), фрукты с фруктами(Персики армянские - Персики азербайджанские), ягоды с ягодами, орехи с орехами.


Можно взять топ-30 по встречаемости пар и закончить на этом наш анализ. Но будет ли это означать, что эти продукты покупатели действительно часто берут вместе (в паре), или же это значит лишь просто то, что первый и/или второй продукт очень популярны, и часто попадают в корзину потребителя друг с другом (при этом эти товары не матчатся друг с другом, как например, Арбуз и Огурцы Луховицкие - третья пара по встречаемости, или Укроп-Арбуз). Если не рассматривать такой вариант событий, то можно просто взять топ-15 самых популярных товаров в магазине и разместить их на одной полке, что приведет скорее к убыткам магазинов, чем если бы самые популярные товары лежали бы вдалеке друг от друга, но при этом рядом с ними лежали бы сопутствующие товары, которые действительно могут идти в паре с другими. Поэтому есть смысл поискать интересные паттерны, выкинув из датафрейма самые популярные продукты, или посмотреть комбинации популярных продуктов и непопулярных.


In [33]:
# Посмотрим, сколько раз купили каждый продукт (число_покупок), и посчитаем процент покупок (percent), в которых этот 
# продукт присутствует, чтобы оценить "популярность" каждого продукта:

products_df = data["Товар"].value_counts().to_frame()
products_df = products_df.rename(columns={"Товар": "число_покупок"})
products_df["percent"] = round(
    data["Товар"].value_counts().to_frame().mul(100) / data["id"].nunique(), 2
)
products_df

число_покупок  percent
Огурцы Луховицкие                 1022    31.23
Арбуз                              978    29.88
Укроп                              828    25.30
Бананы                             691    21.11
Кабачки                            662    20.23
...                                ...      ...
Лук-резанец (шнитт-лук)            124     3.79
Барбарис сушеный                   121     3.70
Капуста белокочанная               121     3.70
Яблоки сушеные                     119     3.64
Абрикос вяленый                    111     3.39

[199 rows x 2 columns]

In [55]:
# Например, самый популярный продукт - Огурцы Луховицкие - есть в 31% покупок или ~ в каждой третьей покупке! Встречаемость
# данного товара в паре практически с любым другим продуктом будет высокой. 

data3[(data3['1_Товар'] == 'Огурцы Луховицкие') | (data3['2_Товар'] == 'Огурцы Луховицкие')]

1_Товар            2_Товар  Встречаемость
2514  Огурцы Луховицкие              Укроп            431
2517  Огурцы Луховицкие              Арбуз            345
2529  Огурцы Луховицкие            Кабачки            326
2511  Огурцы Луховицкие           Петрушка            286
600        Лук репчатый  Огурцы Луховицкие            285
...                 ...                ...            ...
2612  Огурцы Луховицкие       Рычал-Су, 1л             47
2597  Огурцы Луховицкие      Открытка Love             47
2616  Огурцы Луховицкие     Яблоки сушеные             43
2603  Огурцы Луховицкие           Рамбутан             41
2611  Огурцы Луховицкие    Абрикос вяленый             37

[198 rows x 3 columns]

In [64]:
# Поэтому есть смысл рассмотреть комбинации из не самых популярных продуктов, н-р, не из топ-15.
# Список 15 самых популярных продуктов (эти товары встречаются более чем в 15% покупок):

list_of_tops = list(data['Товар'].value_counts().index)[:15]
list_of_tops

['Огурцы Луховицкие',
 'Арбуз',
 'Укроп',
 'Бананы',
 'Кабачки',
 'Лук репчатый',
 'Персики азербайджанские',
 'Баклажаны грунтовые',
 'Персики армянские',
 'Петрушка',
 'Помидоры Парадайз',
 'Кинза',
 'Абрикосы молдавские',
 'Лук зеленый',
 'Кукуруза свежая']

In [65]:
# Датафрейм с товарами не из топ-15 по популярности:
data4 = data3[~data3['1_Товар'].isin(list_of_tops) & ~data3['2_Товар'].isin(list_of_tops)]

# При этом будем рассматривать только пары товаров, встречаемость которых выше 95% пар - такую встречаемость будем считать 
# высокой,т.е. будем искать паттерны только среди 5% самых часто встречающихся пар:

data4 = data4[data4['Встречаемость'] > data3['Встречаемость'].quantile(0.95)]
data4

1_Товар             2_Товар  Встречаемость
3064    Черешня отборная      Черника свежая             87
14                 Лимон   Огурцы Белоруссия             81
4                  Лимон     Морковь немытая             80
56                 Лимон      Чеснок молодой             76
989      Морковь немытая   Огурцы Белоруссия             76
106                Лимон                Мята             75
8533       Морковь мытая      Чеснок молодой             74
7                  Лимон     Капуста цветная             70
11342    Миндаль жареный      Фундук жареный             70
2973   Огурцы Белоруссия               Редис             70
2906   Огурцы Белоруссия      Чеснок молодой             69
14130      Миндаль сырой        Фундук сырой             68
20                 Лимон  Апельсины столовые             68
48                 Лимон       Морковь мытая             68
11402    Миндаль жареный       Кешью жареный             68
982      Морковь немытая     Капуста цветная             64
8622       Морковь мытая           Нектарины             64
123                Лимон               Редис             63
2898   Огурцы Белоруссия       Морковь мытая             62
3088    Черешня отборная      Чеснок молодой             62
196                Лимон    Авокадо стандарт             62
14059      Миндаль сырой         Кешью сырой             61
15                 Лимон    Черешня отборная             61
1604     Капуста цветная      Чеснок молодой             61
2995   Огурцы Белоруссия           Нектарины             60
2865   Огурцы Белоруссия    Черешня отборная             60
9714      Чеснок молодой              Щавель             58
2891   Огурцы Белоруссия   Персики узбекские             58
18                 Лимон       Грибы лисички             57
2988   Огурцы Белоруссия    Грибы шампиньоны             57
2882   Огурцы Белоруссия      Черника свежая             57
3177    Черешня отборная           Нектарины             57
1562     Капуста цветная   Огурцы Белоруссия             57
1026     Морковь немытая              Шпинат             56
163                Лимон       Имбирь корень             56
1101     Морковь немытая              Свекла             56
1031     Морковь немытая      Чеснок молодой             56
8582       Морковь мытая             Цуккини             55
2948   Огурцы Белоруссия      Огурцы Миринда             55
1596     Капуста цветная       Морковь мытая             55
9727      Чеснок молодой     Базилик зеленый             55
98                 Лимон      Огурцы Миринда             55
94                 Лимон     Базилик зеленый             54
9719      Чеснок молодой     Базилик красный             54
86                 Лимон     Базилик красный             54
1056     Морковь немытая              Щавель             54
3084    Черешня отборная       Вишня садовая             53
81                 Лимон              Щавель             53
2936   Огурцы Белоруссия     Базилик красный             52
145                Лимон           Нектарины             52
6029      Черника свежая      Чеснок молодой             52
1098     Морковь немытая               Редис             52

В этом датафрейме можно заметить более интересные паттерны, н-р, Черешня-Черника/Вишня, Лимон-Чеснок/Мята/Корень Имбиря/Авокадо,  Чеснок-Базилик, Миндаль-Фундук-Кешью (жареные\сырые), Морковь-Свекла/Шпинат/Щавель и т.д.
Можно также увеличить список исключенных популярных продуктов до 20 и получить еще пару популярных пар: Капуста цветная - Морковь, Морковь - Цуккини, Черешня - Нектарины.